File: backend-example-java/src/main/java/com/example/backend/adapter/driven/UserPersistenceAdapter.java
package com.example.backend.adapter.driven;

import com.example.backend.application.UserDto;
import com.example.backend.application.UserQueryCriteria;
import com.example.backend.application.UserQueryHandler;
import com.example.backend.application.UserQuerySort;
import com.example.backend.domain.User;
import com.example.backend.domain.UserRepository;
import com.example.backend.domain.UserUniqueSpecification;
import com.example.backend.utility.CryptoUtility;
import com.example.backend.utility.ValueUtility;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
import org.springframework.jdbc.support.rowset.SqlRowSet;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * 用户持久化 Adapter
 */
@Component
public class UserPersistenceAdapter implements UserRepository, UserUniqueSpecification, UserQueryHandler {

    private final SQLiteAdapter sqliteAdapter;
    private final JdbcTemplate jdbcTemplate;
    private final NamedParameterJdbcTemplate namedParameterJdbcTemplate;
    private final String tableName;

    public UserPersistenceAdapter(SQLiteAdapter sqLiteAdapter) {
        this.sqliteAdapter = sqLiteAdapter;
        this.jdbcTemplate = sqLiteAdapter.getJdbcTemplate();
        this.namedParameterJdbcTemplate = sqLiteAdapter.getNamedParameterJdbcTemplate();
        // 每次启动生成唯一表名
        this.tableName = "t_user_" + LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMddHHmmss"));
        // 创建表及默认管理员
        this.jdbcTemplate.execute("""
                CREATE TABLE IF NOT EXISTS %s
                (
                    u_id TEXT PRIMARY KEY,
                    u_username TEXT,
                    u_password TEXT,
                    u_nickname TEXT,
                    u_mobile TEXT,
                    u_is_admin INTEGER,
                    u_created_at TEXT,
                    u_updated_at TEXT
                )
                """.formatted(this.tableName)
        );
        this.jdbcTemplate.execute(
                "DELETE FROM %s WHERE LOWER(u_username) = 'admin'".formatted(this.tableName)
        );
        this.jdbcTemplate.update("""
                        INSERT INTO %s (
                            u_id, u_username, u_password, u_nickname, u_mobile, u_is_admin, u_created_at, u_updated_at
                        ) VALUES (
                            '0', 'admin', ?, '管理员', '', 1, DATETIME('now', 'localtime'), DATETIME('now', 'localtime')
                        )
                        """.formatted(this.tableName),
                CryptoUtility.md5Encode("password")
        );
    }

    //==================================================================================================================
    // UserRepository

    /**
     * 转换成 Entity
     */
    private User toEntity(SqlRowSet sqlRowSet) {
        return User.restore(
                sqlRowSet.getString("u_id"),
                sqlRowSet.getString("u_username"),
                sqlRowSet.getString("u_password"),
                sqlRowSet.getString("u_nickname"),
                sqlRowSet.getString("u_mobile"),
                sqlRowSet.getBoolean("u_is_admin"),
                ValueUtility.toDateTimeOrDefault(sqlRowSet.getString("u_created_at"), LocalDateTime.MIN),
                ValueUtility.toDateTimeOrDefault(sqlRowSet.getString("u_updated_at"), LocalDateTime.MIN)
        );
    }

    @Override
    public void insert(User entity) {
        this.jdbcTemplate.update("""
                        INSERT INTO %s (
                            u_id, u_username, u_password, u_nickname, u_mobile, u_is_admin, u_created_at, u_updated_at
                        ) VALUES (
                            ?, ?, ?, ?, ?, ?, ?, ?
                        )
                        """.formatted(this.tableName),
                entity.getId(),
                entity.getUsername(),
                entity.getPassword(),
                entity.getNickname(),
                entity.getMobile(),
                entity.isAdmin(),
                ValueUtility.formatDateTime(entity.getCreatedAt()),
                ValueUtility.formatDateTime(entity.getUpdatedAt())
        );
    }

    @Override
    public void update(User entity) {
        this.jdbcTemplate.update("""
                        UPDATE %s
                            SET
                                u_username = ?,
                                u_password = ?,
                                u_nickname = ?,
                                u_mobile = ?,
                                u_is_admin = ?,
                                u_created_at = ?,
                                u_updated_at = ?
                            WHERE
                                u_id = ?
                        """.formatted(this.tableName),
                entity.getUsername(),
                entity.getPassword(),
                entity.getNickname(),
                entity.getMobile(),
                entity.isAdmin(),
                ValueUtility.formatDateTime(entity.getCreatedAt()),
                ValueUtility.formatDateTime(entity.getUpdatedAt()),
                entity.getId()
        );
    }

    @Override
    public void deleteById(String id) {
        if (ValueUtility.isEmptyString(id))
            return;
        this.jdbcTemplate.update(
                "DELETE FROM %s WHERE u_id = ?".formatted(this.tableName), id
        );
    }

    @Override
    public User selectById(String id) {
        if (ValueUtility.isEmptyString(id))
            return null;
        SqlRowSet sqlRowSet = this.jdbcTemplate.queryForRowSet(
                "SELECT * FROM %s WHERE u_id = ?".formatted(this.tableName), id
        );
        if (sqlRowSet.next()) {
            return toEntity(sqlRowSet);
        }
        return null;
    }

    @Override
    public User selectByIdReq(String id) {
        User entity = selectById(id);
        if (entity == null)
            throw new PersistenceException(String.format("用户 %s 不存在", id));
        return entity;
    }

    @Override
    public List<User> selectByIds(List<String> ids) {
        List<User> list = new ArrayList<>();
        if (ids == null || ids.isEmpty())
            return list;

        Map<String, Object> params = new HashMap<>();
        params.put("ids", ids);
        SqlRowSet sqlRowSet = this.namedParameterJdbcTemplate.queryForRowSet(
                "SELECT * FROM %s WHERE u_id IN (:ids)".formatted(this.tableName), params
        );
        while (sqlRowSet.next()) {
            User entity = toEntity(sqlRowSet);
            list.add(entity);
        }
        return list;
    }

    @Override
    public User selectByUsername(String username) {
        if (ValueUtility.isEmptyString(username))
            return null;
        SqlRowSet sqlRowSet = this.jdbcTemplate.queryForRowSet(
                "SELECT * FROM %s WHERE LOWER(u_username) = LOWER(?)".formatted(this.tableName), username
        );
        if (sqlRowSet.next()) {
            return toEntity(sqlRowSet);
        }
        return null;
    }

    //==================================================================================================================
    // UserUniqueSpecification

    @Override
    public boolean isUsernameUnique(String username) {
        if (ValueUtility.isEmptyString(username))
            throw new PersistenceException("参数 username 不能为空");
        Integer count = this.jdbcTemplate.queryForObject(
                "SELECT COUNT(*) FROM %s WHERE LOWER(u_username) = LOWER(?)".formatted(this.tableName), int.class, username
        );
        return Integer.valueOf(0).equals(count);
    }

    @Override
    public boolean isNicknameUnique(String nickname) {
        if (ValueUtility.isEmptyString(nickname))
            throw new PersistenceException("参数 nickname 不能为空");
        Integer count = this.jdbcTemplate.queryForObject(
                "SELECT COUNT(*) FROM %s WHERE LOWER(u_nickname) = LOWER(?)".formatted(this.tableName), int.class, nickname
        );
        return Integer.valueOf(0).equals(count);
    }

    @Override
    public boolean isMobileUnique(String mobile) {
        if (ValueUtility.isEmptyString(mobile))
            throw new PersistenceException("参数 mobile 不能为空");
        Integer count = this.jdbcTemplate.queryForObject(
                "SELECT COUNT(*) FROM %s WHERE LOWER(u_mobile) = LOWER(?)".formatted(this.tableName), int.class, mobile
        );
        return Integer.valueOf(0).equals(count);
    }

    //==================================================================================================================
    // UserQueryHandler

    /**
     * 转换成 DTO
     */
    private UserDto toDto(SqlRowSet sqlRowSet) {
        return new UserDto(
                sqlRowSet.getString("u_id"),
                sqlRowSet.getString("u_username"),
                // sqlRowSet.getString("u_password"),
                sqlRowSet.getString("u_nickname"),
                sqlRowSet.getString("u_mobile"),
                sqlRowSet.getBoolean("u_is_admin"),
                ValueUtility.toDateTimeOrDefault(sqlRowSet.getString("u_created_at"), LocalDateTime.MIN),
                ValueUtility.toDateTimeOrDefault(sqlRowSet.getString("u_updated_at"), LocalDateTime.MIN)
        );
    }

    /**
     * 构造查询 SQL
     */
    private String buildQueryCriteria(UserQueryCriteria criteria, Map<String, Object> params) {
        if (criteria == null)
            return "";

        List<String> sqls = new ArrayList<>();
        if (!ValueUtility.isEmptyString(criteria.getKeyword())) {
            sqls.add("u_username LIKE :keyword ESCAPE '\\' OR u_nickname LIKE :keyword ESCAPE '\\'");
            params.put("keyword", this.sqliteAdapter.EscapeLikePattern(criteria.getKeyword()));
        }

        if (!sqls.isEmpty())
            return " WHERE " + String.join(" AND ", sqls);
        return "";
    }

    /**
     * 构造排序 SQL
     */
    private String buildQuerySort(UserQuerySort... sorts) {
        List<String> sqls = new ArrayList<>();
        for (UserQuerySort s : sorts) {
            switch (s) {
                case CREATED_AT_ASC -> sqls.add("u_created_at ASC");
                case CREATED_AT_DESC -> sqls.add("u_created_at DESC");
                case USERNAME_ASC -> sqls.add("u_username ASC");
                case USERNAME_DESC -> sqls.add("u_username DESC");
            }
        }

        if (sqls.isEmpty())
            sqls.add("u_created_at DESC");

        sqls.add("u_id DESC");
        return " ORDER BY " + String.join(", ", sqls);
    }

    @Override
    public UserDto queryById(String id) {
        if (ValueUtility.isEmptyString(id))
            return null;
        SqlRowSet sqlRowSet = this.jdbcTemplate.queryForRowSet(
                "SELECT * FROM %s WHERE u_id = ?".formatted(this.tableName), id
        );
        if (sqlRowSet.next()) {
            return toDto(sqlRowSet);
        }
        return null;
    }

    @Override
    public UserDto queryByIdReq(String id) {
        UserDto dto = queryById(id);
        if (dto == null)
            throw new PersistenceException(String.format("用户 %s 不存在", id));
        return dto;
    }

    @Override
    public int queryCount(UserQueryCriteria criteria) {
        Map<String, Object> params = new HashMap<>();
        String criteriaSql = buildQueryCriteria(criteria, params);
        Integer count = this.namedParameterJdbcTemplate.queryForObject(
                "SELECT COUNT(*) FROM %s %s".formatted(this.tableName, criteriaSql), params, int.class
        );
        return count == null ? 0 : count;
    }

    @Override
    public List<UserDto> query(UserQueryCriteria criteria, UserQuerySort... sorts) {
        Map<String, Object> params = new HashMap<>();
        String criteriaSql = buildQueryCriteria(criteria, params);
        String sortSql = buildQuerySort(sorts);

        SqlRowSet sqlRowSet = this.namedParameterJdbcTemplate.queryForRowSet(
                "SELECT * FROM %s %s %s".formatted(this.tableName, criteriaSql, sortSql), params
        );
        List<UserDto> list = new ArrayList<>();
        while (sqlRowSet.next()) {
            list.add(toDto(sqlRowSet));
        }
        return list;
    }

    @Override
    public List<UserDto> queryByPage(int pageNum, int pageSize, UserQueryCriteria criteria, UserQuerySort... sorts) {
        Map<String, Object> params = new HashMap<>();
        String criteriaSql = buildQueryCriteria(criteria, params);
        String sortSql = buildQuerySort(sorts);

        params.put("limitCount", pageSize);
        params.put("limitOffset", (pageNum - 1) * pageSize);

        SqlRowSet sqlRowSet = this.namedParameterJdbcTemplate.queryForRowSet(
                "SELECT * FROM %s %s %s LIMIT :limitCount OFFSET :limitOffset"
                        .formatted(this.tableName, criteriaSql, sortSql), params
        );
        List<UserDto> list = new ArrayList<>();
        while (sqlRowSet.next()) {
            list.add(toDto(sqlRowSet));
        }
        return list;
    }
}

File: backend-example-dotnet/src/adapter/driven/UserPersistenceAdapter.cs
using System.Dynamic;

using BackendExample.Application;
using BackendExample.Domain;
using BackendExample.Utility;

using Dapper;

using Microsoft.Data.Sqlite;

namespace BackendExample.Adapter.Driven
{
    /// <summary>
    /// 用户持久化 Adapter
    /// </summary>
    public class UserPersistenceAdapter : UserRepository, UserUniqueSpecification, UserQueryHandler
    {
        private readonly SQLiteAdapter sqliteAdapter;
        private readonly string tableName;

        public UserPersistenceAdapter(SQLiteAdapter sqliteAdapter)
        {
            this.sqliteAdapter = sqliteAdapter;
            this.tableName = "t_user" + DateTime.Now.ToString("yyyyMMddHHmmss");
            using var conn = this.sqliteAdapter.GetConnection();
            conn.Execute($"""
                CREATE TABLE IF NOT EXISTS {this.tableName}
                (
                    u_id TEXT PRIMARY KEY,
                    u_username TEXT,
                    u_password TEXT,
                    u_nickname TEXT,
                    u_mobile TEXT,
                    u_is_admin INTEGER,
                    u_created_at TEXT,
                    u_updated_at TEXT
                )
                """);
            conn.Execute($"DELETE FROM {this.tableName} WHERE LOWER(u_username) = 'admin'");
            conn.Execute($"""
                INSERT INTO {this.tableName} (
                    u_id, u_username, u_password, u_nickname, u_mobile, u_is_admin, u_created_at, u_updated_at
                ) VALUES (
                    '0', 'admin', '{CryptoUtility.Md5Encode("password")}', '管理员', '', 1, DATETIME('now', 'localtime'), DATETIME('now', 'localtime')
                )
                """);
        }

        private User ToEntity(dynamic row)
        {
            return User.Restore(
                row.u_id,
                row.u_username,
                row.u_password,
                row.u_nickname,
                row.u_mobile,
                ValueUtility.ToBoolOrDefault(row.u_is_admin.ToString(), false),
                ValueUtility.ToDateTimeOrDefault(row.u_created_at.ToString(), DateTime.MinValue),
                ValueUtility.ToDateTimeOrDefault(row.u_updated_at.ToString(), DateTime.MinValue)
                );
        }

        public async Task Insert(User entity)
        {
            using var conn = this.sqliteAdapter.GetConnection();
            await conn.ExecuteAsync($"""
                INSERT INTO {this.tableName} (
                    u_id, u_username, u_password, u_nickname, u_mobile, u_is_admin, u_created_at, u_updated_at
                ) VALUES (
                    @id, @username, @password, @nickname, @mobile, @isAdmin, @createdAt, @updatedAt
                )
                """, new
            {
                id = entity.Id,
                username = entity.Username,
                password = entity.Password,
                nickname = entity.Nickname,
                mobile = entity.Mobile,
                isAdmin = entity.IsAdmin ? 1 : 0,
                createdAt = entity.CreatedAt.ToString("yyyy-MM-dd HH:mm:ss"),
                updatedAt = entity.UpdatedAt?.ToString("yyyy-MM-dd HH:mm:ss")
            });
        }

        public async Task Update(User entity)
        {
            using var conn = this.sqliteAdapter.GetConnection();
            await conn.ExecuteAsync($"""
                UPDATE {this.tableName}
                    SET
                        u_username = @username,
                        u_password = @password,
                        u_nickname = @nickname,
                        u_mobile = @mobile,
                        u_is_admin = @isAdmin,
                        u_created_at = @createdAt,
                        u_updated_at = @updatedAt
                WHERE
                    u_id = @id
                """, new
            {
                id = entity.Id,
                username = entity.Username,
                password = entity.Password,
                nickname = entity.Nickname,
                mobile = entity.Mobile,
                isAdmin = entity.IsAdmin ? 1 : 0,
                createdAt = entity.CreatedAt.ToString("yyyy-MM-dd HH:mm:ss"),
                updatedAt = entity.UpdatedAt?.ToString("yyyy-MM-dd HH:mm:ss")
            });
        }

        public async Task DeleteById(string id)
        {
            if (ValueUtility.IsEmptyString(id))
                return;
            using var conn = this.sqliteAdapter.GetConnection();
            await conn.ExecuteAsync($"DELETE FROM {this.tableName} WHERE u_id = @id", new { id });
        }

        public async Task<User?> SelectById(string id)
        {
            if (ValueUtility.IsEmptyString(id))
                return null;
            using var conn = this.sqliteAdapter.GetConnection();
            dynamic? row = await conn.QuerySingleOrDefaultAsync(
                $"SELECT * FROM {this.tableName} WHERE u_id = @id", new { id }
                );
            if (row == null)
                return null;
            return ToEntity(row);
        }

        public async Task<User> SelectByIdReq(string id)
        {
            User? entity = await SelectById(id);
            if (entity == null)
                throw new PersistenceException($"用户 {id} 不存在");
            return entity;
        }

        public async Task<IList<User>> SelectByIds(List<string> ids)
        {
            IList<User> list = new List<User>();
            if (ids == null || ids.Count == 0)
                return list;

            using var conn = this.sqliteAdapter.GetConnection();
            IEnumerable<dynamic> rows = await conn.QueryAsync(
                $"SELECT * FROM {this.tableName} WHERE u_id IN @ids", new { ids }
                );
            foreach (dynamic row in rows)
            {
                User entity = ToEntity(row);
                list.Add(entity);
            }
            return list;
        }

        public async Task<User?> SelectByUsername(string username)
        {
            if (ValueUtility.IsEmptyString(username))
                return null;
            using var conn = this.sqliteAdapter.GetConnection();
            dynamic? row = await conn.QuerySingleOrDefaultAsync(
                $"SELECT * FROM {this.tableName} WHERE u_username = @username", new { username }
                );
            if (row == null)
                return null;
            return ToEntity(row);
        }

        //==============================================================================================================
        // UserUniqueSpecification

        public async Task<bool> IsUsernameUnique(string username)
        {
            if (ValueUtility.IsEmptyString(username))
                throw new PersistenceException("参数 username 不能为空");
            using var conn = this.sqliteAdapter.GetConnection();
            return await conn.ExecuteScalarAsync<int>(
                $"SELECT COUNT(*) FROM {this.tableName} WHERE LOWER(u_username) = LOWER(@username)", new { username }
            ) == 0;
        }

        public async Task<bool> IsNicknameUnique(string nickname)
        {
            if (ValueUtility.IsEmptyString(nickname))
                throw new PersistenceException("参数 nickname 不能为空");
            using var conn = this.sqliteAdapter.GetConnection();
            return await conn.ExecuteScalarAsync<int>(
                $"SELECT COUNT(*) FROM {this.tableName} WHERE LOWER(u_nickname) = LOWER(@nickname)", new { nickname }
                ) == 0;
        }

        public async Task<bool> IsMobileUnique(string mobile)
        {
            if (ValueUtility.IsEmptyString(mobile))
                throw new PersistenceException("参数 mobile 不能为空");
            using var conn = this.sqliteAdapter.GetConnection();
            return await conn.ExecuteScalarAsync<int>(
                $"SELECT COUNT(*) FROM {this.tableName} WHERE LOWER(u_mobile) = LOWER(@mobile)", new { mobile }
            ) == 0;
        }

        //==============================================================================================================
        // UserQueryHandler

        private UserDto ToDto(dynamic row)
        {
            return new UserDto(
                row.u_id,
                row.u_username,
                //row.u_password,
                row.u_nickname,
                row.u_mobile,
                ValueUtility.ToBoolOrDefault(row.u_is_admin.ToString(), false),
                ValueUtility.ToDateTimeOrDefault(row.u_created_at.ToString(), DateTime.MinValue),
                ValueUtility.ToDateTimeOrDefault(row.u_updated_at.ToString(), DateTime.MinValue)
                );
        }

        private string BuildQueryCriteria(UserQueryCriteria? criteria, out dynamic @params)
        {
            @params = new ExpandoObject();

            if (criteria == null)
                return string.Empty;

            IList<string> sqls = new List<string>();
            if (!ValueUtility.IsEmptyString(criteria.Keyword))
            {
                sqls.Add("u_username LIKE @Keyword ESCAPE '\\' OR u_nickname LIKE @keyword ESCAPE '\\'");
                @params.keyword = this.sqliteAdapter.EscapeLikePattern(criteria.Keyword);
            }

            if (sqls.Count > 0)
                return " WHERE " + string.Join(" AND ", sqls);

            return string.Empty;
        }

        private string BuildQuerySort(params UserQuerySort[] sorts)
        {
            IList<string> sqls = new List<string>();
            foreach (UserQuerySort sort in sorts)
            {
                switch (sort)
                {
                    case UserQuerySort.CREATED_AT_ASC:
                        sqls.Add("u_created_at ASC");
                        break;
                    case UserQuerySort.CREATED_AT_DESC:
                        sqls.Add("u_created_at DESC");
                        break;
                    case UserQuerySort.USERNAME_ASC:
                        sqls.Add("u_username ASC");
                        break;
                    case UserQuerySort.USERNAME_DESC:
                        sqls.Add("u_username DESC");
                        break;
                }
            }

            if (sqls.Count == 0)
                sqls.Add("u_created_at DESC");

            return " ORDER BY " + string.Join(", ", sqls);
        }

        public async Task<UserDto?> QueryById(string id)
        {
            if (ValueUtility.IsEmptyString(id))
                return null;
            using var conn = this.sqliteAdapter.GetConnection();
            dynamic? row = await conn.QuerySingleOrDefaultAsync(
                $"SELECT * FROM {this.tableName} WHERE u_id = @id", new { id }
                );
            if (row == null)
                return null;
            return ToDto(row);
        }

        public async Task<UserDto> QueryByIdReq(string id)
        {
            UserDto? dto = await QueryById(id);
            if (dto == null)
                throw new PersistenceException($"用户 {id} 不存在");
            return dto;
        }

        public async Task<int> QueryCount(UserQueryCriteria? criteria)
        {
            dynamic @params;
            string crieriaSql = BuildQueryCriteria(criteria, out @params);
            using var conn = this.sqliteAdapter.GetConnection();
            return await conn.ExecuteScalarAsync<int>(
                $"SELECT COUNT(*) FROM {this.tableName} {crieriaSql}", (object)@params
                );
        }

        public async Task<IList<UserDto>> Query(UserQueryCriteria? criteria, params UserQuerySort[] sorts)
        {
            dynamic @params;
            string crieriaSql = BuildQueryCriteria(criteria, out @params);
            string sortSql = BuildQuerySort(sorts);
            using var conn = this.sqliteAdapter.GetConnection();
            IEnumerable<dynamic> rows = await conn.QueryAsync(
                $"SELECT * FROM {this.tableName} {crieriaSql} {sortSql}", (object)@params
                );
            IList<UserDto> list = new List<UserDto>();
            foreach (dynamic row in rows)
            {
                UserDto dto = ToDto(row);
                list.Add(dto);
            }
            return list;
        }

        public async Task<IList<UserDto>> QueryByPage(int pageNum, int pageSize, UserQueryCriteria? criteria,
            params UserQuerySort[] sorts)
        {
            dynamic @params;
            string crieriaSql = BuildQueryCriteria(criteria, out @params);
            string sortSql = BuildQuerySort(sorts);
            @params.limitCount = pageSize;
            @params.limitOffset = (pageNum - 1) * pageSize;

            using var conn = this.sqliteAdapter.GetConnection();
            IEnumerable<dynamic> rows = await conn.QueryAsync(
                $"SELECT * FROM {this.tableName}{crieriaSql}{sortSql} LIMIT @limitCount OFFSET @limitOffset", (object)@params
                );
            IList<UserDto> list = new List<UserDto>();
            foreach (dynamic row in rows)
            {
                UserDto dto = ToDto(row);
                list.Add(dto);
            }
            return list;
        }
    }
}
